### ISSUE TITLE
SoS notebook as a script editor

### ISSUE THREAD

**Issue by gaow** [vatlab/sos-notebook#71]
I find SoS notebook useful as a centralized code editor. I usually do something like: 

```
%save -f cache/test.R
fit  <- function() {...}
```

Long story short is that I cannot execute scripts from notebook for my application. So I write codes in notebook cells, and export them using cell-level `%save`, to some directory in my `.gitignore`. Then I have one notebook for all scripts. This works well except for when I write a random text script eg YAML or JSON and want to save it. Then `%save` will not work, because the text will be evaluated as SoS codes and report errors. 

I wonder 

1. Can there be a switch in `%save` that bypasses syntax check: just dump whatever there is to a file? Or, is there a specific type of cell/magic I can use? It seems [`writefile` magic as shown on this page](http://nbconvert.readthedocs.io/en/latest/customizing.html) is meant to do this, but it does not work for SoS cell.
2. Is SoS for JupyterLab ready at this moment to use for editing SoS codes, not to run any code?

**Comment by gaow** [comment]
Is this the case that `%save` will also run the cell, not just checking the syntax? For other kernels I get something like this:

![2018-05-05-14-56-39_scrot](https://user-images.githubusercontent.com/917985/39667109-9fa2c6ac-5074-11e8-803a-a1660f0ca1f6.png)

This is not what I'm hoping for with `%save` because it is supposed to make a copy of the cell content, not to "judge" if it has good syntax, or to run it.

**Comment by BoPeng** [comment]
I am not sure if `%save` should stop sos from executing the cell, which is the default behavior for "executing". What is the behavior of other kernels, such as `%write` for ipython?

**Comment by gaow** [comment]
Yes, inconsistency with behavior of similar magic in other kernels is also confusing. `writefile` in Python only saves the file, eg:

![2018-05-05-15-02-06_scrot](https://user-images.githubusercontent.com/917985/39667154-6a5912e8-5075-11e8-981c-34e82af804bd.png)

**Comment by BoPeng** [comment]
OK. Try now.

**Comment by gaow** [comment]
Great, all saved without executing!

I have tried SoS JupyterLab a while ago for editing codes, after syntax highlighting more or less worked. But I switched back to SoS Notebook because I miss the `%toc` behavior that allows me to dump between sections quickly. Is it still the case that SoS JupyterLab is lacking that feature!

**Comment by gaow** [comment]
Oh good. So suppose for now I only need JupyterLab as a script organizer using SoS kernel, to run with `%save` magic, not to run any code or data exchange -- for this purpose should I move to jupyter-sos and report back issues? Is it now possible to run stand-alone piece of codes in it, eg R or Python, but no data exchange?

**Comment by BoPeng** [comment]
The problem with `jupyterlab-sos` is that I am pushing many sos features (side panel, line by line execution) to the upstream and changing the upstream to accommodate some behaviors of sos that could only be implemented by hacking the jupyter frontend. JupyterLab is a great project with rigorous coding style, testing mechanism, review system etc, that means it takes time for my PRs to get in, and I cannot do anything to `jupyterlab-sos` when the upstream PRs are pending.

**Comment by gaow** [comment]
I see .. that sounds like a blocker. 

I just tried it out for some basic features. Actually it seems to work for basic user case like I listed above, and the `%toc` came back after installing the `jupyterlab-toc` extension. See my notebook attached.

[test.ipynb.zip](https://github.com/vatlab/sos-notebook/files/1977082/test.ipynb.zip)

However there are 2 glitches not sure if it is related to upstream PR:

1. When I start a new SoS notebook from Jupyter lab there is no cell-level language drop-down selector menu. But if I open one from existing Jupyter Notebook version, then I get the menu as in my attached notebook, and I was able to use it to switch between kernels. Why doesn't it show up by default?

2. Running every cell manually sequentially works well. But when I click on "Restart and run all" there comes an error message:

![2018-05-05-15-53-30_scrot](https://user-images.githubusercontent.com/917985/39667461-7f19011e-507c-11e8-804f-aff0089a5078.png)

Even though it still seems to be working:

![2018-05-05-15-55-08_scrot](https://user-images.githubusercontent.com/917985/39667472-c0f7478a-507c-11e8-816e-bf8b034497fd.png)

3. It feels a bit slow, presumably because many things do not work and drags it slow.

Otherwise as a code editor and non-communicating multi-language doodle pad, it seems to be good enough to actually use, if 1 & 2 are addressed.
